File: languages/java/QuantumSimulator.java
/*
 * Java: Quantum Circuit Simulator
 * Showcasing OOP, interfaces, streams, and modern Java features
 */
import java.util.*;
import java.util.stream.*;
import java.util.concurrent.*;

public class QuantumSimulator {
    
    // Enum for gate types
    public enum GateType {
        HADAMARD("H"),
        PAULI_X("X"),
        PAULI_Y("Y"),
        PAULI_Z("Z"),
        CNOT("CX"),
        SWAP("SWAP");
        
        private final String symbol;
        
        GateType(String symbol) {
            this.symbol = symbol;
        }
        
        public String getSymbol() {
            return symbol;
        }
    }
    
    // Record for quantum gate (Java 14+)
    public record QuantumGate(GateType type, int target, Optional<Integer> control) {
        public QuantumGate {
            if (target < 0) {
                throw new IllegalArgumentException("Target qubit cannot be negative");
            }
            control.ifPresent(c -> {
                if (c < 0) throw new IllegalArgumentException("Control qubit cannot be negative");
            });
        }
        
        @Override
        public String toString() {
            return String.format("%s(q%d%s)", 
                type.getSymbol(), 
                target, 
                control.map(c -> ", c" + c).orElse(""));
        }
    }
    
    // Quantum circuit class
    public static class QuantumCircuit {
        private final String name;
        private final int numQubits;
        private final List<QuantumGate> gates;
        private Map<String, Integer> results;
        
        public QuantumCircuit(String name, int numQubits) {
            this.name = name;
            this.numQubits = numQubits;
            this.gates = new ArrayList<>();
            this.results = null;
            
            if (numQubits <= 0) {
                throw new IllegalArgumentException("Number of qubits must be positive");
            }
        }
        
        public QuantumCircuit addGate(QuantumGate gate) {
            if (gate.target() >= numQubits) {
                throw new IllegalArgumentException(
                    String.format("Gate target %d exceeds circuit width %d", 
                    gate.target(), numQubits));
            }
            gate.control().ifPresent(c -> {
                if (c >= numQubits) {
                    throw new IllegalArgumentException(
                        String.format("Control qubit %d exceeds circuit width %d", 
                        c, numQubits));
                }
            });
            
            gates.add(gate);
            return this;
        }
        
        public CompletableFuture<Map<String, Integer>> simulateAsync(int shots) {
            return CompletableFuture.supplyAsync(() -> {
                System.out.printf("Simulating %s with %d shots...%n", name, shots);
                
                try {
                    Thread.sleep(100); // Simulate computation
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    throw new RuntimeException("Simulation interrupted", e);
                }
                
                // Generate mock results
                results = generateResults(shots);
                return results;
            });
        }
        
        private Map<String, Integer> generateResults(int shots) {
            Map<String, Integer> results = new HashMap<>();
            int numStates = 1 << numQubits;
            
            Random random = new Random();
            int remaining = shots;
            
            for (int i = 0; i < numStates - 1; i++) {
                String state = String.format("%" + numQubits + "s", 
                    Integer.toBinaryString(i)).replace(' ', '0');
                int count = random.nextInt(remaining / 2);
                results.put(state, count);
                remaining -= count;
            }
            
            String lastState = String.format("%" + numQubits + "s", 
                Integer.toBinaryString(numStates - 1)).replace(' ', '0');
            results.put(lastState, remaining);
            
            return results;
        }
        
        public double calculateEntropy() {
            if (results == null) {
                throw new IllegalStateException("No simulation results available");
            }
            
            int totalShots = results.values().stream()
                .mapToInt(Integer::intValue)
                .sum();
            
            return results.values().stream()
                .mapToDouble(count -> (double) count / totalShots)
                .filter(probability -> probability > 0)
                .map(probability -> -probability * (Math.log(probability) / Math.log(2)))
                .sum();
        }
        
        // Factory method for Bell state
        public static QuantumCircuit createBellState() {
            QuantumCircuit circuit = new QuantumCircuit("Bell State", 2);
            circuit.addGate(new QuantumGate(GateType.HADAMARD, 0, Optional.empty()))
                   .addGate(new QuantumGate(GateType.CNOT, 1, Optional.of(0)));
            return circuit;
        }
        
        // Getters
        public String getName() { return name; }
        public int getNumQubits() { return numQubits; }
        public List<QuantumGate> getGates() { return Collections.unmodifiableList(gates); }
        public Optional<Map<String, Integer>> getResults() { 
            return Optional.ofNullable(results); 
        }
        
        @Override
        public String toString() {
            return String.format("QuantumCircuit{name='%s', qubits=%d, gates=%d}", 
                name, numQubits, gates.size());
        }
    }
    
    // Main demonstration
    public static void main(String[] args) {
        System.out.println("=== Java Quantum Simulator ===\n");
        
        // Create Bell state circuit
        QuantumCircuit bellCircuit = QuantumCircuit.createBellState();
        
        // Add some gates
        bellCircuit.addGate(new QuantumGate(GateType.PAULI_X, 1, Optional.empty()))
                   .addGate(new QuantumGate(GateType.PAULI_Z, 0, Optional.empty()));
        
        System.out.println("Circuit: " + bellCircuit);
        
        // Use streams to display gates
        System.out.println("\nGates:");
        bellCircuit.getGates().stream()
            .map(QuantumGate::toString)
            .forEach(System.out::println);
        
        // Async simulation
        ExecutorService executor = Executors.newSingleThreadExecutor();
        try {
            CompletableFuture<Map<String, Integer>> simulation = bellCircuit.simulateAsync(1000);
            
            simulation.thenAccept(results -> {
                System.out.println("\nSimulation Results:");
                
                // Use streams to process results
                results.entrySet().stream()
                    .filter(entry -> entry.getValue() > 100)
                    .sorted(Map.Entry.<String, Integer>comparingByValue().reversed())
                    .forEach(entry -> 
                        System.out.printf("  |%s⟩: %d counts%n", entry.getKey(), entry.getValue()));
                
                // Calculate entropy
                double entropy = bellCircuit.calculateEntropy();
                System.out.printf("%nCircuit Entropy: %.3f bits%n", entropy);
                
                // Additional analysis using streams
                double avgCount = results.values().stream()
                    .mapToInt(Integer::intValue)
                    .average()
                    .orElse(0.0);
                System.out.printf("Average count per state: %.1f%n", avgCount);
                
            }).get(); // Wait for completion
            
        } catch (Exception e) {
            System.err.println("Error during simulation: " + e.getMessage());
            e.printStackTrace();
        } finally {
            executor.shutdown();
        }
        
        // Pattern matching (Java 17+ preview)
        System.out.println("\nCircuit Analysis:");
        analyzeCircuit(bellCircuit);
        
        System.out.println("\n✅ Java quantum simulation completed!");
    }
    
    // Pattern matching demonstration
    private static void analyzeCircuit(QuantumCircuit circuit) {
        // Switch expression
        String complexity = switch (circuit.getNumQubits()) {
            case 1 -> "Trivial";
            case 2 -> "Simple";
            case 3, 4 -> "Medium";
            default -> "Complex";
        };
        
        System.out.printf("Circuit complexity: %s%n", complexity);
        System.out.printf("Number of gates: %d%n", circuit.getGates().size());
        
        // Record pattern matching (Java 19+ preview)
        long cnotGates = circuit.getGates().stream()
            .filter(gate -> gate.type() == GateType.CNOT)
            .count();
        System.out.printf("CNOT gates: %d%n", cnotGates);
    }
}

File: languages/c/quantum_simulator.c
/*
 * C: Quantum Circuit Simulator
 * Showcasing modern C features, memory management, and performance
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <math.h>
#include <time.h>
#include <complex.h>

// Constants
#define MAX_QUBITS 32
#define MAX_GATES 100
#define PI 3.14159265358979323846

// Quantum Gate Types
typedef enum {
    GATE_H,     // Hadamard
    GATE_X,     // Pauli-X
    GATE_Y,     // Pauli-Y
    GATE_Z,     // Pauli-Z
    GATE_CNOT,  // Controlled-NOT
    GATE_SWAP   // SWAP
} GateType;

// Quantum Gate Structure
typedef struct {
    GateType type;
    int target;
    int control;  // -1 if no control qubit
    double angle; // For rotation gates
} QuantumGate;

// Quantum State Vector
typedef struct {
    int num_qubits;
    int dimension;
    double complex* amplitudes;
} QuantumState;

// Quantum Circuit
typedef struct {
    int num_qubits;
    int num_gates;
    QuantumGate gates[MAX_GATES];
    char name[50];
} QuantumCircuit;

// Function Prototypes
QuantumState* create_state(int num_qubits);
void destroy_state(QuantumState* state);
void initialize_state(QuantumState* state);
void apply_gate(QuantumState* state, const QuantumGate* gate);
void print_state(const QuantumState* state);
QuantumCircuit* create_circuit(int num_qubits, const char* name);
void add_gate(QuantumCircuit* circuit, GateType type, int target, int control);
void simulate_circuit(const QuantumCircuit* circuit, QuantumState* state);
double measure_state(const QuantumState* state, int qubit);

// Create quantum state
QuantumState* create_state(int num_qubits) {
    if (num_qubits > MAX_QUBITS) {
        fprintf(stderr, "Error: Too many qubits (max %d)\n", MAX_QUBITS);
        return NULL;
    }
    
    QuantumState* state = malloc(sizeof(QuantumState));
    if (!state) {
        perror("Failed to allocate state");
        return NULL;
    }
    
    state->num_qubits = num_qubits;
    state->dimension = 1 << num_qubits; // 2^n
    
    state->amplitudes = calloc(state->dimension, sizeof(double complex));
    if (!state->amplitudes) {
        perror("Failed to allocate amplitudes");
        free(state);
        return NULL;
    }
    
    return state;
}

// Destroy quantum state
void destroy_state(QuantumState* state) {
    if (state) {
        free(state->amplitudes);
        free(state);
    }
}

// Initialize state to |0...0>
void initialize_state(QuantumState* state) {
    for (int i = 0; i < state->dimension; i++) {
        state->amplitudes[i] = 0.0;
    }
    state->amplitudes[0] = 1.0 + 0.0 * I;
}

// Print quantum state
void print_state(const QuantumState* state) {
    printf("Quantum State (n=%d):\n", state->num_qubits);
    for (int i = 0; i < state->dimension; i++) {
        double complex amp = state->amplitudes[i];
        double prob = creal(amp * conj(amp));
        if (prob > 1e-10) {
            printf("|%0*b>: amplitude = %.3f%+.3fi, probability = %.3f\n",
                   state->num_qubits, i,
                   creal(amp), cimag(amp), prob);
        }
    }
}

// Apply Hadamard gate
void apply_hadamard(QuantumState* state, int qubit) {
    int shift = 1 << qubit;
    int dim = state->dimension;
    
    for (int i = 0; i < dim; i += 2 * shift) {
        for (int j = 0; j < shift; j++) {
            int idx0 = i + j;
            int idx1 = i + j + shift;
            
            double complex a0 = state->amplitudes[idx0];
            double complex a1 = state->amplitudes[idx1];
            
            state->amplitudes[idx0] = (a0 + a1) / sqrt(2.0);
            state->amplitudes[idx1] = (a0 - a1) / sqrt(2.0);
        }
    }
}

// Apply Pauli-X gate
void apply_pauli_x(QuantumState* state, int qubit) {
    int shift = 1 << qubit;
    int dim = state->dimension;
    
    for (int i = 0; i < dim; i += 2 * shift) {
        for (int j = 0; j < shift; j++) {
            int idx0 = i + j;
            int idx1 = i + j + shift;
            
            double complex temp = state->amplitudes[idx0];
            state->amplitudes[idx0] = state->amplitudes[idx1];
            state->amplitudes[idx1] = temp;
        }
    }
}

// Apply gate to state
void apply_gate(QuantumState* state, const QuantumGate* gate) {
    switch (gate->type) {
        case GATE_H:
            apply_hadamard(state, gate->target);
            break;
        case GATE_X:
            apply_pauli_x(state, gate->target);
            break;
        // Additional gates would be implemented here
        default:
            fprintf(stderr, "Gate type %d not implemented\n", gate->type);
    }
}

// Create quantum circuit
QuantumCircuit* create_circuit(int num_qubits, const char* name) {
    QuantumCircuit* circuit = malloc(sizeof(QuantumCircuit));
    if (!circuit) {
        perror("Failed to allocate circuit");
        return NULL;
    }
    
    circuit->num_qubits = num_qubits;
    circuit->num_gates = 0;
    strncpy(circuit->name, name, sizeof(circuit->name) - 1);
    circuit->name[sizeof(circuit->name) - 1] = '\0';
    
    return circuit;
}

// Add gate to circuit
void add_gate(QuantumCircuit* circuit, GateType type, int target, int control) {
    if (circuit->num_gates >= MAX_GATES) {
        fprintf(stderr, "Error: Too many gates in circuit\n");
        return;
    }
    
    QuantumGate* gate = &circuit->gates[circuit->num_gates++];
    gate->type = type;
    gate->target = target;
    gate->control = control;
    gate->angle = 0.0;
}

// Simulate entire circuit
void simulate_circuit(const QuantumCircuit* circuit, QuantumState* state) {
    for (int i = 0; i < circuit->num_gates; i++) {
        apply_gate(state, &circuit->gates[i]);
    }
}

// Main function
int main() {
    printf("=== C Quantum Simulator ===\n");
    
    // Create Bell state circuit
    QuantumCircuit* bell_circuit = create_circuit(2, "Bell State");
    if (!bell_circuit) return 1;
    
    add_gate(bell_circuit, GATE_H, 0, -1);
    add_gate(bell_circuit, GATE_X, 1, 0);  // CNOT with control=0, target=1
    
    // Create and initialize state
    QuantumState* state = create_state(2);
    if (!state) {
        free(bell_circuit);
        return 1;
    }
    
    initialize_state(state);
    
    printf("\nInitial state:\n");
    print_state(state);
    
    // Simulate circuit
    clock_t start = clock();
    simulate_circuit(bell_circuit, state);
    clock_t end = clock();
    
    printf("\nFinal state after %s circuit:\n", bell_circuit->name);
    print_state(state);
    
    double time_taken = ((double)(end - start)) / CLOCKS_PER_SEC;
    printf("\nSimulation time: %.6f seconds\n", time_taken);
    
    // Cleanup
    destroy_state(state);
    free(bell_circuit);
    
    printf("\n✅ Simulation completed successfully\n");
    return 0;
}
